src/base-implementation.ts

```diff
@@ -0,0 +1,103 @@
+import { Wallet } from '@harmony-js/account'
+import { Contract } from '@harmony-js/contract'
+import { Harmony } from '@harmony-js/core'
+import { BigNumber, logger, constants } from 'ethers'
+import { Logger } from 'ethers/lib/utils'
+import { Base } from './base-token'
+
+export class BaseError extends Error {
+  public readonly type: string
+  public readonly code: number
+  public readonly data: string
+
+  constructor(message: string, type: string, code: number, data: string) {
+    super(message)
+    this.name = BaseError.name
+    this.type = type
+    this.code = code
+    this.data = data
+
+    Error.captureStackTrace(this, this.constructor)
+  }
+}
+
+export class BaseToken extends Base {
+  public contract: Contract
+  public isSignerSet = false
+
+  constructor(address: string, abi: any, private client: Harmony) {
+    super()
+    this.contract = this.client.contracts.createContract(abi, address)
+  }
+
+  async _getBalance(address: string, id?: string): Promise<BigNumber> {
+    if (!address) {
+      throw new Error('You have provide an address')
```

Comment:
nit: `you have to`

Reply:
Oh, you're right, I didn't notice that

Reply:
Oh, you're right, I didn't realize before, thanks!